### PR TITLE
feat: Implement odbc driver and share code between sqlite and odbc driver

### PR DESCRIPTION
**Pull request by ghost**
Allows the bot to use many databases rather than just SQLite.
Addresses #3

### CODE REVIEW

File: tests/test_storage.py [hunk 1]
```diff
@@ -9,42 +9,67 @@
 sys.path.append("ebr_trackerbot")
 sys.path.append("ebr_trackerbot/storage")
 
-if os.path.exists("data.db"):
-    os.unlink("data.db")
+
+def delete_fixtures():
```
Comment (eugene-davis) [inline tests/test_storage.py]: If we must actually write to disk (it would be better to mock, but I will admit its pretty terrible mocking the DB libraries) let's at least place the database file in a temp directory with https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory (it still will need to be cleaned up of course). That way there isn't a risk of bumping into a user created one.

File: .pylintrc [hunk 2]
```diff
@@ -15,7 +15,7 @@ ignore-patterns=
 
 # Python code to execute, usually for sys.path manipulation such as
 # pygtk.require().
-init-hook='import sys; sys.path.append("ebr_trackerbot")'
+init-hook='import sys; sys.path.append("ebr_trackerbot"); sys.path.append("ebr_trackerbot/storage")'
```
Comment (eugene-davis) [inline .pylintrc:18]: Why does the normal append (just for ebr_trackerbot) not work here?

File: ebr_trackerbot/storage/db.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        cursor = conn.cursor()
+        cursor.execute(""" SELECT count(name) FROM sqlite_master WHERE type='table' AND name='tracks' """)
+        if cursor.fetchone()[0] == 0:
+            cursor.execute(
+                "CREATE TABLE IF NOT EXISTS tracks "
+                + "(id integer PRIMARY KEY, user text NOT NULL, test text NOT NULL, "
+                + " expiry timestamp NOT NULL, channel_id text NOT NULL, thread_ts text NOT NULL);"
+            )
+            conn.commit()
+    except error as err:
```
Comment (eugene-davis) [inline ebr_trackerbot/storage/db.py]: Can you catch any errors that conn.cursor() might throw more specifically (I believe this would include if the db file can't be accessed due to permissions)? If that is the step that failed, then this exception will cause an additional error when conn.close() is called.

Comment (eugene-davis) [reply ebr_trackerbot/storage/db.py]: This of course applies to the other instances of it as well :)

File: ebr_trackerbot/storage/db.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            data = {
+                "test": row[1],
+                "expiry": row[2] if isinstance(row[2], str) else pendulum.instance(row[2]).to_iso8601_string(),
+                "channel_id": row[3],
+                "thread_ts": row[4],
+            }
+            result.append(data)
+        return result
+    except error as err:
+        logging.error("Error during loading track for user")
```
Comment (eugene-davis) [inline ebr_trackerbot/storage/db.py]: maybe we should also output the user here?

File: ebr_trackerbot/storage/memory.py [hunk 5]
```diff
@@ -4,59 +4,65 @@
 import logging
```
Comment (eugene-davis) [inline ebr_trackerbot/storage/memory.py]: Just a thought about this whole module - would it be possible to use an in-memory sqlite DB in its place?
https://sqlite.org/inmemorydb.html
https://stackoverflow.com/a/38676728

That would let us remove pretty much all the logic specific to the in-memory storage.

File: setup.py [hunk 6]
```diff
@@ -25,6 +25,7 @@
     "pendulum>=2.0.5,<3.0.0",
     "vault-anyconfig>=0.2.2,<0.3.0",
     "PyYAML>=5.1,<6",
+    "pyodbc==4.0.27",
```
Comment (eugene-davis) [inline setup.py]: Can we give this a range, at a minimum I suspect it could be `pyodbc>=4.0.27,<4.1.0` (and probably `<5.0.0` would be ok and also better.)

File: tests/test_storage.py [hunk 7]
```diff
@@ -9,42 +9,67 @@
...
+
+
+reset_fixtures()
+
+from bot import config
+from state import STATE
+
+config["odbc_connection_string"] = "DRIVER={SQLITE3};DATABASE=data-odbc.db"
 
 import pytest
```
Comment (eugene-davis) [inline tests/test_storage.py]: Why are the import statements so low in the file?

File: tests/test_storage.py [hunk 8]
```diff
@@ -9,42 +9,67 @@
...
 
 
 def test_storage():
-    modules = [memory, sqlite]
+    modules = ["memory"]
 
-    for module in modules:
-        assert len(module.load_all_tracked_tests()) == 0
-        module.save(
+    for module_name in modules:
```
Comment (eugene-davis) [inline tests/test_storage.py]: If there's only the one module that is being tested now, why have the for loop? (or if you want to keep it see if you can move it to be from pytest.mark.parameterize)

File: ebr_trackerbot/storage/sqlite.py [hunk 9]
```diff
@@ -3,9 +3,10 @@
 """
```
Comment (eugene-davis) [inline ebr_trackerbot/storage/sqlite.py:3]: Just replace this with the odbc as discussed :+1:

File: ebr_trackerbot/storage/odbc.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    """
+    Retrieve odbc connection
+    """
+    if config.get("odbc_connection_string") is None:
+        raise RuntimeError("Missing odbc_connection_string in configuration")
+    return pyodbc.connect(config.get("odbc_connection_string"))
+
+
+CONN = get_connection()
+db.create_table(get_connection(), pyodbc.DatabaseError)
```
Comment (eugene-davis) [inline ebr_trackerbot/storage/odbc.py]: We should probably handle this by having an init function that can be explicitly called by the CI.
If it makes sense to you, maybe we could have it automatically create this for sqlite (which seems to be common practice) but in a well configured database your user typically won't have the permissions required to create a new table. (this also suggests we need to document, and maybe even print out in the CLI for init what tables are required, and with what permissions for the user).

File: setup.py [hunk 11]
```diff
@@ -25,6 +25,7 @@
     "pendulum>=2.0.5,<3.0.0",
```
Comment (eugene-davis) [inline setup.py:25]: We should also have the dependencies for the various DB possibilities as optional installs. See https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

**Review by eugene-davis — CHANGES_REQUESTED** [review]

**Comment by eugene-davis** [comment]
In regards to the question of how we can test the different drivers in integration tests, one option could be to do it directly in a dockerfile. See https://docs.microsoft.com/en-us/azure/devops/pipelines/languages/docker?view=azure-devops#build-and-test-in-your-dockerfile

This does have the downside of loosing some of the reporting info that Azure Pipelines provides, so it probably shouldn't be first resort.

One other option might be to provide a list of packages that your project needs installed in the images executing tests.

**Comment by eugene-davis** [comment]
Ok, I guess it is worth noting that it might make sense to keep the separate SQLite module, just because it doesn't require an extra C-library installed as it is provided by Python.

File: .pylintrc (hunk 2 above)
Comment (ghost) [reply .pylintrc:18]: I'm not sure but without that python complains that it can not find db.py

Comment (eugene-davis) [reply .pylintrc:18]: Odd, but ok.

File: tests/test_storage.py (hunk 1 above)
Comment (ghost) [reply tests/test_storage.py]: Now it's using tempfile.gettempdir()

File: ebr_trackerbot/storage/odbc.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        get_connection.link = create_connection(config)
+    try:
+        cursor = get_connection.link.cursor()
+        cursor.close()
+    except:
+        get_connection.link = create_connection(config)
+    return get_connection.link
+
+
+get_connection.link = None
```
Comment (martinm82) [inline ebr_trackerbot/storage/odbc.py:32]: Where is this object coming from?

File: tests/integration/test_storage.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""Tests for `ebr_trackerbot` package."""
+
+import sys
+import os
+
+sys.path.append("ebr_trackerbot")
```
Comment (martinm82) [inline tests/integration/test_storage.py]: Why is this needed? During a test run the module get's installed

File: tests/integration/test_storage.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+from bot import config
+from state import STATE
+import pytest
+import sqlite
+import odbc
+import tempfile
+
+
+def delete_fixtures():
```
Comment (martinm82) [inline tests/integration/test_storage.py:12]: Use the py.test fixtures instead: https://docs.pytest.org/en/latest/fixture.html?highlight=teardown#fixtures-as-function-arguments.

Will make the code cleaner

File: tests/integration/test_storage.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+def reset_fixtures():
+    delete_fixtures()
+    open(tempfile.gettempdir() + "/data-odbc.db", "a").close()
+
+
+def test_storage():
+    reset_fixtures()
+    config["odbc_connection_string"] = "DRIVER={SQLITE3};DATABASE=" + tempfile.gettempdir() + "/data-odbc.db"
+    modules = ["odbc", "sqlite"]
```
Comment (martinm82) [inline tests/integration/test_storage.py]: Maybe instead it should be a parameterized test?

File: tests/unit/test_storage_db.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""Tests for `ebr_trackerbot` package."""
+
+import sys
+import os
+
+sys.path.append("ebr_trackerbot")
```
Comment (martinm82) [inline tests/unit/test_storage_db.py]: This should not be necessary. See my other comment

File: tests/unit/test_storage_db.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+import db
+from unittest.mock import Mock
+
+
+def test_storage_db_create_table():
+    conn = Mock()
+    executeMock = Mock()
+    conn.cursor.return_value = executeMock
+
+    db.create_table(lambda: conn)
```
Comment (martinm82) [inline tests/unit/test_storage_db.py]: I don't think this test is necessary as it is testing the original Python module itself which is (or should be) already tested

File: ebr_trackerbot/storage/db.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+        cursor = conn.cursor()
+        cursor.execute(
+            "CREATE TABLE IF NOT EXISTS tracks "
+            + "(id integer PRIMARY KEY, user text NOT NULL, test text NOT NULL, "
+            + " expiry timestamp NOT NULL, channel_id text NOT NULL, thread_ts text NOT NULL);"
+        )
+        conn.commit()
+    except:
+        logging.error(str(sys.exc_info()[1]))
+        conn.close()
```
Comment (martinm82) [inline ebr_trackerbot/storage/db.py]: Close should be called in my opinion also in case the execution succeeded

File: ebr_trackerbot/storage/db.py [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    """
+    result = []
+    conn = get_connection()
+    try:
+        cursor = conn.cursor()
+        cursor.execute(""" SELECT user, test, expiry, channel_id, thread_ts FROM tracks; """)
+        rows = cursor.fetchall()
+        for row in rows:
+            data = {
+                "user": row[0],
```
Comment (martinm82) [inline ebr_trackerbot/storage/db.py:87]: I think the cursor returns a named tuple so you can use here user, test, etc. instead indices.
Makes your code safe against accidental order changes in the select

File: ebr_trackerbot/storage/odbc.py (hunk 12 above)
Comment (ghost) [reply ebr_trackerbot/storage/odbc.py:32]: Which object?

File: ebr_trackerbot/storage/db.py (hunk 19 above)
Comment (ghost) [reply ebr_trackerbot/storage/db.py:87]: I double checked it and fetchall returns only list no named tuple :(

File: ebr_trackerbot/storage/db.py (hunk 18 above)
Comment (ghost) [reply ebr_trackerbot/storage/db.py]: Why it should re-connect to database after every query?

File: tests/unit/test_storage_db.py (hunk 17 above)
Comment (ghost) [reply tests/unit/test_storage_db.py]: This unit test checks that custom function create_table calls specific sql command

File: ebr_trackerbot/storage/db.py [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+    try:
+        cursor = conn.cursor()
+        cursor.execute(
+            "CREATE TABLE IF NOT EXISTS tracks "
+            + "(id integer PRIMARY KEY, user text NOT NULL, test text NOT NULL, "
+            + " expiry timestamp NOT NULL, channel_id text NOT NULL, thread_ts text NOT NULL);"
+        )
+        conn.commit()
+    except:
+        logging.error(str(sys.exc_info()[1]))
```
Comment (eugene-davis) [inline ebr_trackerbot/storage/db.py]: I don't think that having the program continue if this call fails is really a use-case.
I would instead propose changing this section to a `with` statement, if both libraries will support it with the same syntax (I'm not sure that the SQLite one will).
https://github.com/mkleehammer/pyodbc/wiki/Objects#exception-handling

File: ebr_trackerbot/storage/db.py (hunk 18 above)
Comment (eugene-davis) [reply ebr_trackerbot/storage/db.py]: @martinm82 I think instead that close should not be called here at all, as the connection is not actually created here.

File: ebr_trackerbot/storage/odbc.py [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+def create_connection(config):
+    if config.get("odbc_connection_string") is None:
+        raise RuntimeError("Missing odbc_connection_string in configuration")
+    link = pyodbc.connect(config.get("odbc_connection_string"))
+    db.create_table(lambda: link)
+    return link
+
+
+def get_connection(config):
```
Comment (eugene-davis) [inline ebr_trackerbot/storage/odbc.py]: I think this is unnecessary, as it looks like connection pooling is built-in.
https://github.com/mkleehammer/pyodbc/wiki/Module#pooling

File: tests/integration/test_storage.py (hunk 13 above)
Comment (ghost) [reply tests/integration/test_storage.py]: I rewrote the way how to import modules

File: tests/integration/test_storage.py (hunk 14 above)
Comment (ghost) [reply tests/integration/test_storage.py:12]: The name is probably misleading. I don't need to pass some fixtures I need to prepare some files so I implemented setup and teardown functions.

File: tests/integration/test_storage.py (hunk 15 above)
Comment (ghost) [reply tests/integration/test_storage.py]: Ok

File: tests/unit/test_storage_db.py (hunk 16 above)
Comment (ghost) [reply tests/unit/test_storage_db.py]: Fixed

File: ebr_trackerbot/storage/db.py (hunk 20 above)
Comment (ghost) [reply ebr_trackerbot/storage/db.py]: ok. I rewrote it

File: tests/unit/common.py [hunk 22]
```diff
@@ -0,0 +1,17 @@
+"""
+Common functions for tests
+"""
+
+
+def get_payload(post_message):
```
Comment (eugene-davis) [inline tests/unit/common.py]: I would say this belongs in a fixture. It can be added to conftest.py to allow sharing.
https://docs.pytest.org/en/latest/fixture.html#scope-sharing-a-fixture-instance-across-tests-in-a-class-module-or-session

File: setup.py (hunk 6 above)
Comment (ghost) [reply setup.py]: ok

File: tests/integration/test_storage.py (hunk 14 above)
Comment (eugene-davis) [reply tests/integration/test_storage.py:12]: @bestdeveloperever I agree with @martinm82 here, because you can do a cleaner setup and teardown via fixtures. Take a look at https://docs.pytest.org/en/latest/fixture.html#fixture-finalization-executing-teardown-code
It should allow condensing these first three functions into a single function.

File: ebr_trackerbot/storage/odbc.py (hunk 12 above)
Comment (martinm82) [reply ebr_trackerbot/storage/odbc.py:32]: get_connection

File: tests/integration/test_storage.py (hunk 14 above)
Comment (martinm82) [reply tests/integration/test_storage.py:12]: But you can use py.test fixtures to achieve this kind of stuff. Imagine your test will create hundreds of file handles and throw an exception. The teardown won't be called to cleanup them

File: ebr_trackerbot/storage/odbc.py (hunk 12 above)
Comment (ghost) [reply ebr_trackerbot/storage/odbc.py:32]: it's a static variable for get_connection function.

File: tests/integration/test_storage.py (hunk 14 above)
Comment (ghost) [reply tests/integration/test_storage.py:12]: but I see in documentation:
For tearDown(): “This method will only be called if the setUp() succeeds, regardless of the outcome of the test method.”
So if test throw exception then teardown function should be called

**Review by eugene-davis — CHANGES_REQUESTED** [review]

File: ebr_trackerbot/storage/db.py (hunk 19 above)
Comment (martinm82) [reply ebr_trackerbot/storage/db.py:87]: Ok, it is indeed no named tuple but a list. The docs say something different: https://github.com/mkleehammer/pyodbc/wiki/Cursor#fetchall

And based on the code it says you can use column names to access the column values: https://github.com/mkleehammer/pyodbc/blob/master/src/cursor.h#L152

File: ebr_trackerbot/storage/db.py (hunk 3 above)
Comment (ghost) [reply ebr_trackerbot/storage/db.py]: done

File: ebr_trackerbot/storage/memory.py (hunk 5 above)
Comment (ghost) [reply ebr_trackerbot/storage/memory.py]: Memory storage was removed

File: tests/test_storage.py (hunk 7 above)
Comment (ghost) [reply tests/test_storage.py]: The module was rewritten

File: ebr_trackerbot/storage/odbc.py (hunk 10 above)
Comment (ghost) [reply ebr_trackerbot/storage/odbc.py]: This was rewritten

File: tests/test_storage.py (hunk 8 above)
Comment (ghost) [reply tests/test_storage.py]: It was rewritten

File: ebr_trackerbot/storage/db.py (hunk 18 above)
Comment (ghost) [reply ebr_trackerbot/storage/db.py]: It was rewritten with "with" statement

File: tests/unit/common.py (hunk 22 above)
Comment (ghost) [reply tests/unit/common.py]: ok

File: ebr_trackerbot/storage/db.py (hunk 4 above)
Comment (ghost) [reply ebr_trackerbot/storage/db.py]: This part was already removed

File: tests/integration/test_storage.py (hunk 14 above)
Comment (ghost) [reply tests/integration/test_storage.py:12]: This is integration test

**Review by eugene-davis — APPROVED** [review]

File: tests/integration/test_storage.py (hunk 14 above)
Comment (martinm82) [reply tests/integration/test_storage.py:12]: @bestdeveloperever you are referencing documentation from the unittest module but not py.test.
Right now what you do is a home-brew solution to do a setup and teardown. As @eugene-davis pointed out, there is a more elegant way of doing it.

Comment (martinm82) [reply tests/integration/test_storage.py:12]: To correct what I was saying before. I was still having your initial version in mind when saying "homebrew". So the new version is better, but there is even a more elegant one.
